### PR TITLE
Refactoring monitor channel

### PR DESCRIPTION
**Pull request by tayvs**
Simplify monitorChannel and fix potential bug

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by tayvs** [comment]
@raboof plz review

**Comment by raboof** [comment]
OK TO TEST

**Comment by tayvs** [comment]
@raboof it seems that CI pipeline is green But it not displayed in github((

File: runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala [hunk 1]
```diff
@@ -88,43 +89,31 @@ object ChannelUtils {
...
-                  monitor(connectionAttempts)
-                }
-              }
-            })
-        }
+        channel.notifyWhenStateChanged(currentState, new Runnable {
+          def run(): Unit = channel.getState(false) match {
+            case fail @ ConnectivityState.TRANSIENT_FAILURE => monitor(fail, connectionAttempts + 1)
+            case ready @ ConnectivityState.READY            => monitor(ready, 0)
+            case ConnectivityState.SHUTDOWN                 => done.trySuccess(Done)
```
Comment (raboof) [inline runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala]: Moving this inside the `notifyWhenStateChanged` seems dangerous when the initial state might already be `SHUTDOWN`. That makes me wonder:
* would it be possible to do the `match` outside of `notifyWhenStateChanged` entirely?
* is it necessary to call `notifyWhenStateChanged` on each recursive call, or would once be sufficient?

Comment (tayvs) [reply runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala]: @raboof If initial state `SHUTDOWN` I complete promise outside `monitor` function and not go into it.
We must to call `notifyWhenStateChanged` each time to track state changes. It call callback only once when passed state would not equal to current anymore.
I think that moving matching outside the `notifyWhenStateChanged` function can be done. I'll try to do it

Comment (tayvs) [reply runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala]: @raboof If initial state `SHUTDOWN` I complete promise outside `monitor` function and not go into it.
We must to call `notifyWhenStateChanged` each time to track state changes. It call callback only once when passed state would not equal to current anymore.
I think that moving matching outside the `notifyWhenStateChanged` function can be done. I'll try to do it

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by tayvs** [comment]
@raboof What do you think about new variant?

**Comment by raboof** [comment]
Yes, that looks more straightforward to me!

**Comment by tayvs** [comment]
> Yes, that looks more straightforward to me!

Ok, So I'll delete other variants and rebase with master and try to test in real service

**Comment by raboof** [comment]
Great!

File: runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala [hunk 2]
```diff
@@ -52,32 +52,35 @@ object ChannelUtils {
...
-      if (maxConnectionAttempts.contains(connectionAttempts)) {
-        // shutdown is idempotent in ManagedChannelImpl
-        channel.shutdown()
-        done.tryFailure(new ClientConnectionException(s"Unable to establish connection after [$maxConnectionAttempts]"))
-      } else {
-        val currentState = channel.getState(false)
-        if (currentState == ConnectivityState.SHUTDOWN) {
+    def monitor(currentState: ConnectivityState, connectionAttempts: Int): Unit = {
+      //TODO: remove before merge
+      println(s"monitoring with state $currentState and connectionAttempts $connectionAttempts")
```
Comment (raboof) [inline runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala]: Remove this? :)

Comment (raboof) [reply runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala]: > I want to test and debug new monitoring before.

Cool, I'll wait with merging then - looking forward to it!

Comment (tayvs) [reply runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala]: @raboof Good news. Now netty heartbeat triggers `TRANSIENT_FAILURE`! Regarding `println` I think that it good idea to have such logging. What do you think about it and what kind of logger should I use?

Comment (tayvs) [reply runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala]: @raboof Is there any ETA of 0.7.4 release?

File: runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala [hunk 3]
```diff
@@ -52,32 +52,35 @@ object ChannelUtils {
...
-                  // IDLE / CONNECTING / SHUTDOWN
-                  monitor(currentState, connectionAttempts)
-                }
-            })
-        }
-      }
+          None
 
+        // IDLE / CONNECTING
+        case _ =>
```
Comment (ennru) [inline runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala]: You might as well spell those two out here.

File: runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala (hunk 2 above)
Comment (ennru) [reply runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala]: We're aiming for 1.0.0 now.

Comment (raboof) [reply runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala]: > I think that it good idea to have such logging.

That makes sense. For now I think it's OK to pass the `LoggingAdapter` available in `ClientState` on to here via the `NettyClientUtils`.

> Is there any ETA of 0.7.4 release?

We don't have one planned specifically, but I think we can release one soon. Perhaps we should also try and get #811 tested and included - have you seen that one? would you be interested in testing it?

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

File: runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala (hunk 2 above)
Comment (tayvs) [reply runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala]: @raboof I think I can test #811 . Config are same and I just need to build project with that branch, am I right?

File: runtime/src/test/scala/akka/grpc/internal/ChannelUtilsSpec.scala [hunk 4]
```diff
@@ -10,6 +10,9 @@ import akka.Done
 import akka.grpc.internal.ChannelUtilsSpec.FakeChannel
 import io.grpc._
 import ConnectivityState._
+import akka.event.{ Logging, LoggingAdapter, NoLogging }
+import akka.http.impl.util.DefaultNoLogging
+import com.sun.xml.internal.ws.resources.ClientMessages
```
Comment (raboof) [inline runtime/src/test/scala/akka/grpc/internal/ChannelUtilsSpec.scala]:
```suggestion
```

File: runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala (hunk 2 above)
Comment (raboof) [reply runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala]: Yes - you'll need to set `akka.grpc.client."*".grpc-load-balancing = "round_robin"`

**Comment by tayvs** [comment]
@raboof @ennru it seems that I broke binary compatibility. Is it Ok? Should I add filters?

**Comment by raboof** [comment]
> it seems that I broke binary compatibility. Is it Ok? Should I add filters?

Since we're still before 1.0.0 I think it's OK to add filters :+1: . CONTRIBUTING.md still talks about a 'src/main/mima-filters/<last-released-version>.backwards.excludes file', but I think it's neater to create it as a directory and put the filters in a file in that directory.

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

File: runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala (hunk 2 above)
Comment (raboof) [reply runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala]: #811 has now been merged and a snapshot was released as version 0.7.3+72-96e15516 - do you think you'll have time to look into testing that?

**Pull request merged**